homeassistant/components/tplink/deprecate.py

```diff
@@ -0,0 +1,120 @@
+"""Helper class for deprecating entities."""
+
+from __future__ import annotations
+
+from collections.abc import Sequence
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+from homeassistant.components.automation import automations_with_entity
+from homeassistant.components.script import scripts_with_entity
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import entity_registry as er
+from homeassistant.helpers.issue_registry import IssueSeverity, async_create_issue
+
+from .const import DOMAIN
+
+if TYPE_CHECKING:
+    from .entity import CoordinatedTPLinkFeatureEntity, TPLinkFeatureEntityDescription
+
+
+@dataclass(slots=True)
+class DeprecatedInfo:
+    """Class to define deprecation info for deprecated entities."""
+
+    platform: str
+    new_platform: str
+    breaks_in_ha_version: str
+
+
+def async_check_create_deprecated(
+    hass: HomeAssistant,
+    unique_id: str,
+    entity_description: TPLinkFeatureEntityDescription,
+) -> bool:
+    """Return true if the entity should be created based on the deprecated_info.
+
+    If deprecated_info is not defined will return true.
+    If entity not yet created will return false.
+    If entity disabled will return false.
+    """
+    if not entity_description.deprecated_info:
+        return True
+
+    deprecated_info = entity_description.deprecated_info
+    platform = deprecated_info.platform
+
+    ent_reg = er.async_get(hass)
+    entity_id = ent_reg.async_get_entity_id(
+        platform,
+        DOMAIN,
+        unique_id,
+    )
+    if not entity_id:
+        return False
+
+    entity_entry = ent_reg.async_get(entity_id)
+    assert entity_entry
+    if entity_entry.disabled:
+        return False
+
+    return True
+
+
+def async_cleanup_deprecated(
+    hass: HomeAssistant,
+    platform: str,
+    entry_id: str,
+    entities: Sequence[CoordinatedTPLinkFeatureEntity],
+) -> None:
+    """Remove deprecated entities and .
+
+    If deprecated_info is not defined will return true.
+    If entity not yet created will return false.
+    If entity disabled will delete it and return false.
+    Otherwise will return true and create issues for scripts or automations.
+    """
+    ent_reg = er.async_get(hass)
+    for entity in entities:
+        if not (deprecated_info := entity.entity_description.deprecated_info):
+            continue
```

Comment:
Currently all buttons are being deprecated so this won't be covered. I could add the ` async_cleanup_deprecated` to all the platforms?

Reply:
Its fine to either add it to all platforms or make a new function that calls both `CoordinatedTPLinkFeatureEntity.entities_for_device_and_its_children` and `async_cleanup_deprecated` as long as `entities_for_device_and_its_children` doesn't have the unexpected side effects